spec.bs

```diff
@@ -6,57 +6,417 @@ Status: CG-DRAFT
+    1. Let |url_list| be an empty {{list}}.
+    1. For each |url_with_metadata| in |urls|:
+        1. If |url_with_metadata| has no field "|url|", reject |config_promise|.
+        1. Otherwise, let |url_string| be |url_with_metadata|["|url|"].
+        1. Let |url| be the result of running a <a href="https://url.spec.whatwg.org/#concept-url-parser">URL parser</a> on |url_string|.
+        1. If |url| is not a valid {{URL}}, reject |config_promise|.
+        1. Otherwise, [=list/append=] |url_string| to |url_list|.
+        1. If |url_with_metadata| has field "|reportingMetadata|":
+            1. Let |reporting_metadata| be |url_with_metadata|["|reportingMetadata|"].
+            1. If |reporting_metadata| is invalid, reject |config_promise|.
+    1. <a href="https://html.spec.whatwg.org/multipage/infrastructure.html#enqueue-the-following-steps">Enqueue the following steps</a>:
+        1. If {{Worklet/addModule()}} has not yet been called, reject |config_promise|.
+        1. Let |pending_url| be a newly generated <a href="https://www.rfc-editor.org/rfc/rfc4122">urn:uuid</a>.
+        1. Let |fenced_frame_config| be a <a href="https://wicg.github.io/fenced-frame/fencedframeconfig">FencedFrameConfig</a> constructed from |pending_url|.
+        1. Resolve |config_promise| as |fenced_frame_config|.
+        1. Let |index_promise| be a new promise.
+        1. Let |worklet_queue| be the <a href="https://html.spec.whatwg.org/multipage/infrastructure.html#parallel-queue">parallel queue</a> for the context's associated {{SharedStorageWorklet}} interface, if it exists.
+        1. If |worklet_queue| does not exist, create it.
+        1. <a href="https://html.spec.whatwg.org/multipage/infrastructure.html#enqueue-the-following-steps">Enqueue the following steps</a> on |worklet_queue|            1. Let |index| be 0.
+            1. If module script has been loaded for this worklet:
+                1. Let |operation_map| be the operation map in the associated {{SharedStorageWorkletGlobalScope}}.
+                1. Let |operation| be the operation registered to |name| in |operation_map|, if |operation| exists.
+                1. If |operation| exists:
+                    1. If |options| contains |data|, let |operation_result| be the result of invoking |operation| with arguments |url_list| and |data|.
+                    1. Otherwise, let |operation_result| be the result of invoking |operation| with argument |url_list|.
+                    1. If there are no errors:
+                        1. Set |index| to the result of casting |operation_result| to an {{unsigned long}}. If this throws an exception, catch it and set |index| back to 0.
+                        1. If |index| > |url_list|.[=list/size=], then set |index| back to 0.
+            1. Resolve |index_promise| with |index|.
+        1. When |index_promise| is resolved, let |result_index| be its numerical value.
```

Comment:
This sounds like there's a wait. But in reality (as in the current implementation), the JS should start allowing other operation to execute. So I think we need to queue a task, which runs the steps:
- Wait until |index_promise| is resolved.
- ....

I'm not sure if my suggestion is ideal. So perhaps leave the comment as is and wait for an spec expert.

Reply:
I believe you have to use the `[=upon fulfillment=]` hook here to take action when a promise resolves:

https://webidl.spec.whatwg.org/#upon-fulfillment

Something like `[=Upon fulfillment=] of |index_promise| perform the following steps:`.

Reply:
Done.